R/geom-defaults.r

```diff
@@ -7,14 +7,42 @@
 #' @keywords internal
 #' @export
 #' @examples
-#' update_geom_defaults("point", list(colour = "darkblue"))
+#'
+#' # updating a geom's default aesthetic settings
+#' # example: change geom_point()'s default color
+#' GeomPoint$default_aes
+#' update_geom_defaults("point", list(color = "red"))
+#' GeomPoint$default_aes
 #' ggplot(mtcars, aes(mpg, wt)) + geom_point()
+#'
+#' # reset default
 #' update_geom_defaults("point", list(colour = "black"))
+#'
+#' # note: adding default aesthetics is possible
+#' # update_geom_defaults("point", list(foo = "bar"))
+#' # GeomPoint$default_aes
+#'
+#'
+#' # updating a stat's default aesthetic settings
+#' # example: change stat_bin()'s default y-axis to the density scale
+#' StatBin$default_aes
+#' update_stat_defaults("bin", list(y = expr(after_stat(density))))
+#' StatBin$default_aes
+#' ggplot(data.frame(x = rnorm(1e3)), aes(x)) +
+#'   geom_histogram() +
+#'   geom_function(fun = dnorm, color = "red")
```

Comment:
I like this example as I've always wondered myself what `update_stat_defaults()` might be useful for. I was surprised this doesn't work on the CRAN version, so that is an extra bonus. If you want to get rid of the `expr()` you could also just use `update_stat_defaults("bin", aes(y = after_stat(density)))`, which might be somewhat more explicit.

Reply:
That doesn't quite work:
``` r
# without expr()
update_stat_defaults("bin", list(y = after_stat(density)))
StatBin$default_aes
#> Aesthetic mapping: 
#> * `x`      -> `after_stat(count)`
#> * `y`      -> `function (x, ...) 
#> UseMethod("density")`
#> * `weight` -> 1

StatBin$default_aes |> str()
#> List of 3
#>  $ x     : language ~after_stat(count)
#>   ..- attr(*, ".Environment")=<environment: namespace:ggplot2> 
#>  $ y     :function (x, ...)  
#>  $ weight: num 1
#>  - attr(*, "class")= chr "uneval"

ggplot(data.frame(x = rnorm(1e3)), aes(x)) +
  geom_histogram() +
  geom_function(fun = dnorm, color = "red")
#> Error in `strip_dots()` at ggplot2/R/aes-evaluation.r:330:4:
#> ! Unknown input: <function>
```

Interestingly, the environment of the promise is wrong with my `expr()` version, but it works:
``` r
# with expr()
update_stat_defaults("bin", list(y = expr(after_stat(density))))
StatBin$default_aes
#> Aesthetic mapping: 
#> * `x`      -> `after_stat(count)`
#> * `y`      -> `after_stat(density)`
#> * `weight` -> 1

StatBin$default_aes |> str()
#> List of 3
#>  $ x     : language ~after_stat(count)
#>   ..- attr(*, ".Environment")=<environment: namespace:ggplot2> 
#>  $ y     : language ~after_stat(density)
#>   ..- attr(*, ".Environment")=<environment: R_GlobalEnv> 
#>  $ weight: num 1
#>  - attr(*, "class")= chr "uneval"

ggplot(data.frame(x = rnorm(1e3)), aes(x)) +
  geom_histogram() +
  geom_function(fun = dnorm, color = "red")
#> `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
```

![](https://i.imgur.com/TtiIkco.png)<!-- -->


The environment attached to the promise seems not to matter. Example:
``` r
density <- 0
after_stat(density) # evaluated in .GlobalEnv
#> [1] 0

ggplot(data.frame(x = rnorm(1e3)), aes(x)) +
  geom_histogram() +
  geom_function(fun = dnorm, color = "red")
#> `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
```

![](https://i.imgur.com/RS0Cha4.png)<!-- -->

My takeaway:
1. Just writing `after_stat(density)`, without `expr()`, evaluates it in `.GlobalEnv`. If it's not set, `density` defaults to `stats::density`, the function itself, which of course breaks.
2. With `expr()` passes the expression along, but with the wrong environment. (It's dynamically using the environment in which the assignment is made.) But clearly **ggplot2** ignores that environment, so it works. 

Thoughts?

Reply:
Sorry I have failed to attract attention to the salient bit: instead of using `list()`, one can use `aes()` to capture the expression with the environment, and that should work fine as input for the updater.

Reply:
Oh, wow! Yep, totally missed that part! Pushing now...